File: app/src/main/java/com/hippo/ehviewer/ui/scene/ServerListScene.java
package com.hippo.ehviewer.ui.scene;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.hippo.ehviewer.EhApplication;
import com.hippo.ehviewer.ServiceRegistry;
import com.hippo.ehviewer.EhDB;
import com.hippo.ehviewer.R;
import com.hippo.ehviewer.client.lrr.LRRApiUtilsKt;
import com.hippo.ehviewer.client.lrr.LRRAuthManager;
import com.hippo.ehviewer.client.lrr.LRRUrlHelper;
import com.hippo.ehviewer.client.lrr.data.LRRServerInfo;
import com.hippo.ehviewer.dao.ServerProfile;
import com.hippo.ehviewer.ui.scene.gallery.list.GalleryListScene;
import com.hippo.scene.Announcer;
import com.hippo.util.IoThreadPoolExecutor;

import okhttp3.OkHttpClient;

import java.util.ArrayList;
import java.util.List;

/**
 * Server list management scene. Shows all saved server profiles in a list.
 * Tap to switch, long-press for edit/delete, FAB to add new server.
 */
public final class ServerListScene extends BaseScene {

    private RecyclerView mRecyclerView;
    private TextView mEmptyText;
    private ServerAdapter mAdapter;
    private List<ServerProfile> mProfiles = new ArrayList<>();

    @Override
    public boolean needShowLeftDrawer() {
        return true;
    }

    @Nullable
    @Override
    public View onCreateView2(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                              @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.scene_server_list, container, false);

        MaterialToolbar toolbar = view.findViewById(R.id.toolbar);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());

        mRecyclerView = view.findViewById(R.id.recycler_view);
        mEmptyText = view.findViewById(R.id.empty_text);
        FloatingActionButton fab = view.findViewById(R.id.fab_add);

        mRecyclerView.setLayoutManager(new LinearLayoutManager(getEHContext()));
        mAdapter = new ServerAdapter();
        mRecyclerView.setAdapter(mAdapter);

        fab.setOnClickListener(v -> showAddDialog());

        loadProfiles();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
        // Refresh list when returning from ServerConfigScene
        loadProfiles();
    }

    private void loadProfiles() {
        Context ctx = getEHContext();
        if (ctx == null) return;

        com.hippo.util.IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
            java.util.List<com.hippo.ehviewer.dao.ServerProfile> result =
                new ArrayList<>(EhDB.getAllServerProfiles());
            result.sort((a, b) -> Boolean.compare(b.isActive(), a.isActive()));
            requireActivity().runOnUiThread(() -> {
                mProfiles = result;
                if (mAdapter != null) mAdapter.notifyDataSetChanged();
                if (mEmptyText != null) {
                    mEmptyText.setVisibility(mProfiles.isEmpty() ? View.VISIBLE : View.GONE);
                }
            });
        });
    }

    private void switchToProfile(ServerProfile profile) {
        Context ctx = getEHContext();
        if (ctx == null) return;

        // DB writes on IO thread, then UI updates on main thread
        com.hippo.util.IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
            EhDB.deactivateAllProfiles();
            EhDB.updateServerProfile(new ServerProfile(
                    profile.getId(), profile.getName(), profile.getUrl(),
                    null, true));
            android.app.Activity a = getActivity();
            if (a != null) a.runOnUiThread(() -> switchToProfileUiUpdate(profile));
        });
    }

    private void switchToProfileUiUpdate(ServerProfile profile) {
        Context ctx = getEHContext();
        if (ctx == null) return;

        // Update LRRAuthManager
        LRRAuthManager.setServerUrl(profile.getUrl());
        String profileApiKey = LRRAuthManager.getApiKeyForProfile(profile.getId());
        LRRAuthManager.setApiKey(profileApiKey);
        LRRAuthManager.setServerName(profile.getName());
        LRRAuthManager.setActiveProfileId(profile.getId());

        // Clear caches so previous server's content doesn't appear under new server
        ServiceRegistry.INSTANCE.clearAllCaches();

        // Reload download manager on IO thread (EhDB.getAllDownloadInfo blocks)
        IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
            ServiceRegistry.INSTANCE.getDataModule().getDownloadManager().reload();
        });

        Toast.makeText(ctx, getString(R.string.lrr_server_switched, profile.getName()),
                Toast.LENGTH_SHORT).show();

        // Warn if switching to HTTP on a public address
        String url = profile.getUrl();
        if (url != null && url.toLowerCase().startsWith("http://")
                && !LRRUrlHelper.isLanAddress(url)) {
            Toast.makeText(ctx, R.string.lrr_security_warning, Toast.LENGTH_LONG).show();
        }

        // Navigate to archive list
        Bundle args = new Bundle();
        args.putString(GalleryListScene.KEY_ACTION, GalleryListScene.ACTION_HOMEPAGE);
        if (getActivity() instanceof com.hippo.scene.StageActivity) {
            ((com.hippo.scene.StageActivity) getActivity())
                    .startSceneFirstly(new Announcer(GalleryListScene.class).setArgs(args));
        }
    }

    private void deleteProfile(ServerProfile profile, int position) {
        Context ctx = getEHContext();
        if (ctx == null) return;

        new AlertDialog.Builder(ctx)
                .setTitle(R.string.lrr_delete_server)
                .setMessage(getString(R.string.lrr_delete_server_confirm, profile.getName()))
                .setPositiveButton(android.R.string.ok, (d, w) -> {
                    LRRAuthManager.clearApiKeyForProfile(profile.getId());
                    IoThreadPoolExecutor.Companion.getInstance().execute(() ->
                        EhDB.deleteServerProfile(profile));
                    mProfiles.remove(position);
                    mAdapter.notifyItemRemoved(position);
                    if (mEmptyText != null) {
                        mEmptyText.setVisibility(mProfiles.isEmpty() ? View.VISIBLE : View.GONE);
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    // ===== Adapter =====

    private class ServerAdapter extends RecyclerView.Adapter<ServerViewHolder> {

        @NonNull
        @Override
        public ServerViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_server_profile, parent, false);
            return new ServerViewHolder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull ServerViewHolder holder, int position) {
            ServerProfile profile = mProfiles.get(position);
            holder.name.setText(profile.getName());
            holder.url.setText(profile.getUrl());

            // Highlight the currently active/connected server
            if (profile.isActive()) {
                holder.activeIcon.setVisibility(View.VISIBLE);
                holder.name.setTextColor(holder.itemView.getContext().getResources().getColor(R.color.colorPrimary));
                holder.itemView.setBackgroundColor(Color.parseColor("#1A4CAF50")); // subtle green tint
            } else {
                holder.activeIcon.setVisibility(View.INVISIBLE);
                holder.name.setTextColor(Color.parseColor("#999999"));
                holder.url.setTextColor(Color.parseColor("#BBBBBB"));
                holder.itemView.setBackgroundColor(Color.TRANSPARENT);
            }

            holder.itemView.setOnClickListener(v -> switchToProfile(profile));
            holder.itemView.setOnLongClickListener(v -> {
                showProfileOptions(profile, position);
                return true;
            });
        }

        @Override
        public int getItemCount() {
            return mProfiles.size();
        }
    }

    private static class ServerViewHolder extends RecyclerView.ViewHolder {
        final TextView name;
        final TextView url;
        final ImageView activeIcon;

        ServerViewHolder(@NonNull View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.server_name);
            url = itemView.findViewById(R.id.server_url);
            activeIcon = itemView.findViewById(R.id.icon_active);
        }
    }

    private void showProfileOptions(ServerProfile profile, int position) {
        Context ctx = getEHContext();
        if (ctx == null) return;

        String[] options = {
                getString(R.string.lrr_edit_server),
                getString(R.string.lrr_delete_server)
        };

        new AlertDialog.Builder(ctx)
                .setTitle(profile.getName())
                .setItems(options, (d, which) -> {
                    if (which == 0) {
                        showEditDialog(profile, position);
                    } else if (which == 1) {
                        deleteProfile(profile, position);
                    }
                })
                .show();
    }

    private void showEditDialog(ServerProfile profile, int position) {
        Context ctx = getEHContext();
        if (ctx == null) return;

        View dialogView = LayoutInflater.from(ctx).inflate(R.layout.dialog_edit_server, null);
        com.google.android.material.textfield.TextInputEditText nameEdit =
                dialogView.findViewById(R.id.edit_server_name);
        com.google.android.material.textfield.TextInputEditText urlEdit =
                dialogView.findViewById(R.id.edit_server_url);
        com.google.android.material.textfield.TextInputEditText apiKeyEdit =
                dialogView.findViewById(R.id.edit_api_key);

        // Pre-fill current values
        nameEdit.setText(profile.getName());
        urlEdit.setText(profile.getUrl());
        String existingKey = LRRAuthManager.getApiKeyForProfile(profile.getId());
        if (existingKey != null) {
            apiKeyEdit.setText(existingKey);
        }

        AlertDialog dialog = new AlertDialog.Builder(ctx)
                .setTitle(R.string.lrr_edit_server)
                .setView(dialogView)
                .setPositiveButton(R.string.lrr_edit_server_save, null) // set below to prevent auto-dismiss
                .setNegativeButton(android.R.string.cancel, null)
                .create();

        // Prevent the dialog's keyboard from resizing the underlying Activity layout.
        // Save the original mode so we can restore it when the dialog is dismissed.
        final int originalSoftInputMode;
        if (getActivity() != null && getActivity().getWindow() != null) {
            originalSoftInputMode = getActivity().getWindow().getAttributes().softInputMode;
            getActivity().getWindow().setSoftInputMode(
                    WindowManager.LayoutParams.SOFT_INPUT_ADJUST_NOTHING);
        } else {
            originalSoftInputMode = WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN;
        }

        dialog.setOnDismissListener(d -> {
            // Restore the Activity's original soft input mode
            if (getActivity() != null && getActivity().getWindow() != null) {
                getActivity().getWindow().setSoftInputMode(originalSoftInputMode);
            }
        });

        dialog.show();

        // Override positive button to add validation
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(v -> {
            String newName = nameEdit.getText() != null ? nameEdit.getText().toString().trim() : "";
            String newUrl = urlEdit.getText() != null ? urlEdit.getText().toString().trim() : "";
            String newKey = apiKeyEdit.getText() != null ? apiKeyEdit.getText().toString().trim() : "";

            if (newName.isEmpty()) {
                nameEdit.setError(getString(R.string.name_is_empty));
                return;
            }
            if (newUrl.isEmpty()) {
                urlEdit.setError(getString(R.string.lrr_server_url_empty));
                return;
            }

            // Normalize URL: add protocol if missing, strip trailing slashes
            String normalizedUrl = LRRUrlHelper.normalizeUrl(newUrl);
            if (!LRRUrlHelper.hasExplicitScheme(normalizedUrl)) {
                // Default to https:// for bare hostnames
                normalizedUrl = "https://" + normalizedUrl;
            }

            // Save to DB on IO thread
            final ServerProfile updated = new ServerProfile(
                    profile.getId(), newName, normalizedUrl,
                    null, profile.isActive());
            IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
                EhDB.updateServerProfile(updated);
                LRRAuthManager.setApiKeyForProfile(profile.getId(), newKey.isEmpty() ? null : newKey);
                if (profile.isActive()) {
                    LRRAuthManager.setServerUrl(updated.getUrl());
                    LRRAuthManager.setApiKey(newKey.isEmpty() ? null : newKey);
                    LRRAuthManager.setServerName(newName);
                }
            });

            // Refresh list immediately with in-memory data
            mProfiles.set(position, updated);
            mAdapter.notifyItemChanged(position);
            dialog.dismiss();
        });
    }

    /**
     * Show a dialog to add a new server profile.
     * Reuses dialog_edit_server.xml layout. Tests connection with HTTPS→HTTP
     * fallback, then saves and switches to the new server.
     */
    private void showAddDialog() {
        Context ctx = getEHContext();
        if (ctx == null) return;

        View dialogView = LayoutInflater.from(ctx).inflate(R.layout.dialog_edit_server, null);
        com.google.android.material.textfield.TextInputEditText nameEdit =
                dialogView.findViewById(R.id.edit_server_name);
        com.google.android.material.textfield.TextInputEditText urlEdit =
                dialogView.findViewById(R.id.edit_server_url);
        com.google.android.material.textfield.TextInputEditText apiKeyEdit =
                dialogView.findViewById(R.id.edit_api_key);

        AlertDialog dialog = new AlertDialog.Builder(ctx)
                .setTitle(R.string.lrr_add_server)
                .setView(dialogView)
                .setPositiveButton(R.string.lrr_save_and_connect, null)
                .setNegativeButton(android.R.string.cancel, null)
                .create();

        dialog.show();

        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(v -> {
            String name = nameEdit.getText() != null ? nameEdit.getText().toString().trim() : "";
            String url = urlEdit.getText() != null ? urlEdit.getText().toString().trim() : "";
            String apiKey = apiKeyEdit.getText() != null ? apiKeyEdit.getText().toString().trim() : "";

            if (name.isEmpty()) {
                nameEdit.setError(getString(R.string.name_is_empty));
                return;
            }
            if (url.isEmpty()) {
                urlEdit.setError(getString(R.string.lrr_server_url_empty));
                return;
            }

            // Normalize: trim + strip trailing slashes
            final String normalizedInput = LRRUrlHelper.normalizeUrl(url);
            final String finalKey = apiKey.isEmpty() ? null : apiKey;

            // Disable button during connection test
            dialog.getButton(AlertDialog.BUTTON_POSITIVE).setEnabled(false);
            Toast.makeText(ctx, R.string.lrr_test_connection, Toast.LENGTH_SHORT).show();

            // Temporarily set auth for the interceptor
            String oldUrl = LRRAuthManager.getServerUrl();
            String oldKey = LRRAuthManager.getApiKey();
            LRRAuthManager.setApiKey(finalKey);

            // Build a test client with short timeouts
            OkHttpClient baseClient = ServiceRegistry.INSTANCE.getNetworkModule().getOkHttpClient();
            OkHttpClient testClient = LRRUrlHelper.buildTestClient(baseClient);

            IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
                // Use HTTPS→HTTP fallback logic
                LRRUrlHelper.connectWithFallback(testClient, normalizedInput,
                        new LRRUrlHelper.ConnectCallback() {
                    @Override
                    public void onSuccess(@NonNull String resolvedUrl,
                                          @NonNull LRRServerInfo info,
                                          boolean usedHttpFallback) {
                        if (getActivity() == null) return;
                        getActivity().runOnUiThread(() -> {
                            // Set auth immediately (in-memory via .apply())
                            LRRAuthManager.setServerUrl(resolvedUrl);
                            LRRAuthManager.setApiKey(finalKey);
                            LRRAuthManager.setServerName(name);

                            // DB operations on IO thread
                            IoThreadPoolExecutor.Companion.getInstance().execute(() -> {
                                EhDB.deactivateAllProfiles();
                                ServerProfile newProfile = new ServerProfile(
                                        0, name, resolvedUrl, null, true);
                                long newId = EhDB.insertServerProfile(newProfile);
                                LRRAuthManager.setApiKeyForProfile(newId, finalKey);
                                LRRAuthManager.setActiveProfileId(newId);
                                ServiceRegistry.INSTANCE.getDataModule().getDownloadManager().reload();
                            });

                            Toast.makeText(ctx, getString(R.string.lrr_connection_success,
                                    info.name, info.version), Toast.LENGTH_SHORT).show();

                            // Warn if connected via HTTP on a public address
                            if (usedHttpFallback && !LRRUrlHelper.isLanAddress(resolvedUrl)) {
                                Toast.makeText(ctx, R.string.lrr_security_warning,
                                        Toast.LENGTH_LONG).show();
                            }

                            dialog.dismiss();
                            loadProfiles();

                            // Navigate to archive list
                            Bundle args = new Bundle();
                            args.putString(GalleryListScene.KEY_ACTION,
                                    GalleryListScene.ACTION_HOMEPAGE);
                            if (getActivity() instanceof com.hippo.scene.StageActivity) {
                                ((com.hippo.scene.StageActivity) getActivity())
                                        .startSceneFirstly(new Announcer(GalleryListScene.class)
                                                .setArgs(args));
                            }
                        });
                    }

                    @Override
                    public void onFailure(@NonNull Exception error) {
                        // Restore old auth on failure
                        LRRAuthManager.setServerUrl(oldUrl);
                        LRRAuthManager.setApiKey(oldKey);

                        if (getActivity() == null) return;
                        getActivity().runOnUiThread(() -> {
                            dialog.getButton(AlertDialog.BUTTON_POSITIVE).setEnabled(true);
                            Toast.makeText(ctx, getString(R.string.lrr_connection_failed,
                                    LRRApiUtilsKt.friendlyError(ctx, error)),
                                    Toast.LENGTH_LONG).show();
                        });
                    }
                });
            });
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mRecyclerView = null;
        mEmptyText = null;
        mAdapter = null;
    }
}

File: app/src/main/java/com/hippo/ehviewer/ui/scene/ServerListScene.kt
package com.hippo.ehviewer.ui.scene

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.hippo.android.resource.AttrResources
import com.hippo.ehviewer.R
import com.hippo.ehviewer.ServiceRegistry
import com.hippo.ehviewer.util.collectFlow
import com.hippo.ehviewer.util.collectFlowWhileCreated
import com.lanraragi.reader.client.api.LRRAuthManager
import com.lanraragi.reader.client.api.LRRSecureStorageUnavailableException
import com.lanraragi.reader.client.api.LRRUrlHelper
import com.lanraragi.reader.client.api.friendlyError
import com.hippo.ehviewer.dao.ServerProfile
import com.hippo.ehviewer.ui.scene.gallery.list.GalleryListScene
import com.hippo.scene.Announcer
import com.hippo.scene.StageActivity

/**
 * Server list management scene. Shows all saved server profiles in a list.
 * Tap to switch, long-press for edit/delete, FAB to add new server.
 *
 * Business logic (CRUD, connection testing, cache clearing) lives in
 * [ServerListViewModel]. Dialog construction lives in [ServerListDialogHelper].
 * This Scene retains: view inflation, adapter/RecyclerView setup, ViewModel
 * observation, and navigation.
 */
class ServerListScene : BaseScene() {

    private lateinit var viewModel: ServerListViewModel
    private lateinit var dialogHelper: ServerListDialogHelper

    private var mRecyclerView: RecyclerView? = null
    private var mEmptyText: TextView? = null
    private var mAdapter: ServerAdapter? = null
    private var mReauthDialogShown = false

    // Local copy for adapter binding; kept in sync with ViewModel's profiles.
    private var mProfiles: List<ServerProfile> = emptyList()

    // Snapshot of the list last dispatched to the adapter. Kept in sync with
    // mProfiles after EVERY notify*() call (full reload via DiffUtil, single
    // remove via notifyItemRemoved, single change via notifyItemChanged).
    // See docs/diffutil-root-cause-analysis.md for the snapshot ownership rule.
    private var mLastSnapshot: List<ServerProfile> = emptyList()

    override fun needShowLeftDrawer(): Boolean = true

    override fun onCreateView2(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        viewModel = ViewModelProvider(requireActivity())[ServerListViewModel::class.java]
        dialogHelper = ServerListDialogHelper(
            contextProvider = { getEHContext() },
            activityProvider = { activity as? androidx.fragment.app.FragmentActivity },
            stringProvider = { resId -> getString(resId) },
            stringFormatProvider = { resId, arg -> getString(resId, arg) },
            viewModel = viewModel
        )

        val view = inflater.inflate(R.layout.scene_server_list, container, false)

        val toolbar = view.findViewById<MaterialToolbar>(R.id.toolbar)
        toolbar.setNavigationOnClickListener { onBackPressed() }

        mRecyclerView = view.findViewById(R.id.recycler_view)
        mEmptyText = view.findViewById(R.id.empty_text)
        val fab = view.findViewById<FloatingActionButton>(R.id.fab_add)

        mRecyclerView?.layoutManager = LinearLayoutManager(getEHContext())
        mAdapter = ServerAdapter()
        mRecyclerView?.adapter = mAdapter

        fab.setOnClickListener { dialogHelper.showAddDialog() }

        viewModel.loadProfiles()
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        observeViewModel()
    }

    override fun onResume() {
        super.onResume()
        // Refresh list when returning from ServerConfigScene
        viewModel.loadProfiles()
        // Prompt user if encrypted keystore is unavailable (credentials lost)
        if (LRRAuthManager.isNeedsReauthentication() && !mReauthDialogShown) {
            mReauthDialogShown = true
            dialogHelper.showReauthDialog()
        }
    }

    // ===== ViewModel observation =====

    private fun observeViewModel() {
        // UI-4: view-scoped + STARTED-gated. The previous fragment-scoped
        // collectors stacked one set per re-attach — applyProfileSwitch
        // (cache clear + reload + navigation) ran once per accumulated
        // collector for a single ProfileActivated event.
        collectFlow(viewLifecycleOwner, viewModel.profiles) { profiles ->
            val adapter = mAdapter
            if (adapter != null) {
                val diff = DiffUtil.calculateDiff(
                    ServerProfileDiffCallback(mLastSnapshot, profiles)
                )
                mProfiles = profiles
                mLastSnapshot = ArrayList(profiles)
                diff.dispatchUpdatesTo(adapter)
            } else {
                mProfiles = profiles
                mLastSnapshot = ArrayList(profiles)
            }
            mEmptyText?.visibility = if (mProfiles.isEmpty()) View.VISIBLE else View.GONE
        }

        // Observe one-shot UI events. These are replay=0 events that drive
        // state-machine transitions (ProfileActivated → applyProfileSwitch,
        // ProfileAdded/EditSaved → navigation/dialog dismissal); losing one in a
        // STOPPED window would split Room's active flag from LRRAuthManager or
        // soft-lock the add/edit dialog, so collect for the whole view lifetime.
        // viewLifecycleOwner scoping still cancels on view destroy, so no
        // duplicate collector stacks across re-attach.
        collectFlowWhileCreated(viewLifecycleOwner, viewModel.uiEvent) { event ->
            handleUiEvent(event)
        }
    }

    private fun handleUiEvent(event: ServerListViewModel.ServerListUiEvent) {
        val ctx = getEHContext() ?: return
        when (event) {
            is ServerListViewModel.ServerListUiEvent.ShowToast -> {
                Toast.makeText(ctx, event.message, Toast.LENGTH_LONG).show()
            }

            is ServerListViewModel.ServerListUiEvent.ShowToastRes -> {
                Toast.makeText(ctx, event.resId, Toast.LENGTH_LONG).show()
            }

            is ServerListViewModel.ServerListUiEvent.SecureStorageError -> {
                dialogHelper.showSecureStorageErrorDialog()
            }

            is ServerListViewModel.ServerListUiEvent.ProfileActivated -> {
                applyProfileSwitch(event.profile)
            }

            is ServerListViewModel.ServerListUiEvent.ProfileAdded -> {
                handleProfileAdded(event)
            }

            is ServerListViewModel.ServerListUiEvent.EditSaved -> {
                dialogHelper.currentEditDialog?.dismiss()
                dialogHelper.clearDialogRefs()
            }

            is ServerListViewModel.ServerListUiEvent.EditConnectionFailed -> {
                dialogHelper.currentEditDialog
                    ?.getButton(androidx.appcompat.app.AlertDialog.BUTTON_POSITIVE)
                    ?.isEnabled = true
                Toast.makeText(
                    ctx,
                    getString(R.string.lrr_connection_failed, friendlyError(ctx, event.cause)),
                    Toast.LENGTH_LONG
                ).show()
            }

            is ServerListViewModel.ServerListUiEvent.AddConnectionFailed -> {
                dialogHelper.currentAddDialog
                    ?.getButton(androidx.appcompat.app.AlertDialog.BUTTON_POSITIVE)
                    ?.isEnabled = true
                Toast.makeText(
                    ctx,
                    getString(R.string.lrr_connection_failed, friendlyError(ctx, event.cause)),
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    // ===== Profile switch (UI side) =====

    private fun applyProfileSwitch(profile: ServerProfile) {
        val ctx = getEHContext() ?: return

        // Update LRRAuthManager
        try {
            LRRAuthManager.setServerUrl(profile.url)
            val profileApiKey = LRRAuthManager.getApiKeyForProfile(profile.id)
            LRRAuthManager.setApiKey(profileApiKey)
            LRRAuthManager.setServerName(profile.name)
            LRRAuthManager.setActiveProfileId(profile.id)
            LRRAuthManager.setAllowCleartext(profile.allowCleartext)
            LRRAuthManager.bumpServerConfigVersion()
        } catch (e: LRRSecureStorageUnavailableException) {
            dialogHelper.showSecureStorageErrorDialog()
            return
        }

        // Clear caches so previous server's content doesn't appear under new server
        ServiceRegistry.clearAllCaches()

        // Reload download manager on main thread (repo.assertMainThread)
        ServiceRegistry.dataModule.downloadManager.reload()

        Toast.makeText(
            ctx,
            getString(R.string.lrr_server_switched, profile.name),
            Toast.LENGTH_SHORT
        ).show()

        // Warn if switching to HTTP on a public address
        if (LRRUrlHelper.isInsecureWanUrl(profile.url)) {
            Toast.makeText(ctx, R.string.lrr_security_warning, Toast.LENGTH_LONG).show()
        }

        navigateToGalleryList()
    }

    // ===== Profile add (UI side) =====

    private fun handleProfileAdded(event: ServerListViewModel.ServerListUiEvent.ProfileAdded) {
        val ctx = getEHContext() ?: return

        // Reload download manager on main thread
        ServiceRegistry.dataModule.downloadManager.reload()

        Toast.makeText(
            ctx,
            getString(R.string.lrr_connection_success, event.info.name, event.info.version),
            Toast.LENGTH_SHORT
        ).show()

        if (event.usedHttpFallback) {
            Toast.makeText(ctx, R.string.lrr_https_fallback_warning, Toast.LENGTH_LONG).show()
        } else if (LRRUrlHelper.isInsecureWanUrl(event.resolvedUrl)) {
            Toast.makeText(ctx, R.string.lrr_security_warning, Toast.LENGTH_LONG).show()
        }

        dialogHelper.currentAddDialog?.dismiss()
        dialogHelper.clearDialogRefs()

        navigateToGalleryList()
    }

    // ===== Navigation =====

    private fun navigateToGalleryList() {
        val args = Bundle().apply {
            putString(GalleryListScene.KEY_ACTION, GalleryListScene.ACTION_HOMEPAGE)
        }
        val act = activity
        if (act is StageActivity) {
            act.startSceneFirstly(Announcer(GalleryListScene::class.java).setArgs(args))
        }
    }

    // ===== Adapter =====

    private inner class ServerAdapter : RecyclerView.Adapter<ServerViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ServerViewHolder {
            val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_server_profile, parent, false)
            return ServerViewHolder(v)
        }

        override fun onBindViewHolder(holder: ServerViewHolder, position: Int) {
            val profile = mProfiles[position]
            holder.name.text = profile.name
            holder.url.text = profile.url

            // Highlight the currently active/connected server. Colors resolve from
            // the ACTIVITY THEME (manual in-app setting), never from -night
            // resources — the manual switcher uses Activity.setTheme, so -night
            // would follow the system dark mode instead (RES-4). Inactive rows are
            // normal list items: name = theme textColorPrimary; the URL keeps the
            // layout's textColorSecondary (no code override) for both branches.
            val ctx = holder.itemView.context
            if (profile.isActive) {
                holder.activeIcon.visibility = View.VISIBLE
                holder.name.setTextColor(
                    AttrResources.getAttrColor(ctx, R.attr.serverActiveNameColor)
                )
                holder.itemView.setBackgroundColor(
                    AttrResources.getAttrColor(ctx, R.attr.serverActiveRowBg)
                )
            } else {
                holder.activeIcon.visibility = View.INVISIBLE
                holder.name.setTextColor(
                    AttrResources.getAttrColor(ctx, android.R.attr.textColorPrimary)
                )
                holder.itemView.setBackgroundColor(Color.TRANSPARENT)
            }

            holder.itemView.setOnClickListener { viewModel.activateProfile(profile) }
            holder.itemView.setOnLongClickListener {
                dialogHelper.showProfileOptions(profile, position) { p, pos ->
                    dialogHelper.showEditDialog(p, pos)
                }
                true
            }
        }

        override fun getItemCount(): Int = mProfiles.size
    }

    private class ServerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val name: TextView = itemView.findViewById(R.id.server_name)
        val url: TextView = itemView.findViewById(R.id.server_url)
        val activeIcon: ImageView = itemView.findViewById(R.id.icon_active)
    }

    // ===== DiffUtil =====

    /**
     * DiffUtil callback for ServerProfile lists. Identity is the Room PK `id`.
     * Content compares all fields rendered in onBindViewHolder so that an edit
     * (rename, URL change, active toggle) repaints the affected row.
     */
    private class ServerProfileDiffCallback(
        private val oldList: List<ServerProfile>,
        private val newList: List<ServerProfile>
    ) : DiffUtil.Callback() {

        override fun getOldListSize(): Int = oldList.size
        override fun getNewListSize(): Int = newList.size

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition].id == newList[newItemPosition].id
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            val o = oldList[oldItemPosition]
            val n = newList[newItemPosition]
            return o.name == n.name && o.url == n.url && o.isActive == n.isActive
        }
    }

    // ===== Lifecycle =====

    override fun onDestroyView() {
        super.onDestroyView()
        mRecyclerView = null
        mEmptyText = null
        mAdapter = null
        mLastSnapshot = emptyList()
        dialogHelper.clearDialogRefs()
    }
}
